rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,368 @@
+ - Built-in parameter provides very limited flexibility (no mask, no runtime).
+
+### The changes:
+
+#### New quantization scheme
+
+```
+(dst - zp_dst) * scale_dst = post_ops(OP(...))
+```
+Notes:
+- scales/zero_points are applied to their respective argument
+  - To get dst output of post_ops() should be divided(!) on scale_dst
+- if a users needs rescaling/reshifting in the post chain, they can rely on binary post-op
+- Scales will be supported for all data types
+
+
+#### A unified scaling API using mechanism `attr::set_scales`
+
+- The encoding of `DNNL_ARG_SCALE` is changed to `512` to avoid overlapping with other arguments.
+- Mapping of existing scaling to `attr::set_scales`:
+  - `attr::output_scales`
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_DST, mask, output_scales);
+    ```
+  - `attr::p_op::binary_mul`
+    ``` cpp
+     primitive_attr attr;
+     attr.set_scales(DNNL_ARG_DST, mask, output_scales);
+    ```
+  - `attr::rnn_data_qparams`
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_SRC, mask, data_scales);
+    attr.set_scales(DNNL_ARG_DST, mask, data_scales);
+    ```
+  - `attr::rnn_weights_qparams`
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_WEIGHTS, mask, weights_scales);
+    ```
+  - `attr::p_op::built-in parameter`
+    ``` cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+    p_ops.append_eltwise(algorithm::eltwise_relu, alpha, beta);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0, /* binary scale */ {DNNL_RUNTIME_F32_VAL});
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(1) | DNNL_ARG_DST,
+            /* mask */ 0, /* eltwise scale */ {DNNL_RUNTIME_F32_VAL});
+
+    ...
+    std::unordered_map<int, memory> args;
+    ...
+    args.insert({DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1 | DNNL_ARG_SCALE, p_op_0_src_1_scale_mem});
+    args.insert({DNNL_ARG_ATTR_MULTIPLE_POST_OP(1) | DNNL_ARG_DST | DNNL_ARG_SCALE, p_op_1_scale_mem});
+    p.execute(s, args);
+    ```
+ - `built-in parameter` in Sum
+   ``` cpp
+   primitive_attr attr;
+   attr.set_scales(DNNL_ARG_MULTIPLE_SRC + 0,
+           /* mask */ 0, {DNNL_RUNTIME_F32_VAL});
+   attr.set_scales(DNNL_ARG_MULTIPLE_SRC + 1,
+           /* mask */ 0, {DNNL_RUNTIME_F32_VAL});
+
+   ...
+   std::unordered_map<int, memory> args;
+   ...
+   args.insert({DNNL_ARG_MULTIPLE_SRC + 0, scale_0_mem});
+   args.insert({DNNL_ARG_MULTIPLE_SRC + 1, scale_1_mem});
+   p.execute(s, args);
+   ```
+
+
+#### Creation-time scaling support will be dropped.
+``` c
+dnnl_status_t DNNL_API dnnl_primitive_attr_get_scales(
+        dnnl_primitive_attr_t attr, int arg, dnnl_dim_t *count, int *mask);
+
+
+dnnl_status_t DNNL_API dnnl_primitive_attr_set_scales(
+        dnnl_primitive_attr_t attr, int arg, dnnl_dim_t count, int mask);
+```
```

Comment:
Might be better to replace with C++ API which will be removed.

Reply:
I will add C++ API